src/super4pcs/algorithms/match4pcsBase.cc

```diff
@@ -0,0 +1,951 @@
+                                             static_cast<Scalar>(kMinNumberOfTrials)));
+    // We use a simple heuristic to elevate the probability to a reasonable value
+    // given that we don't simply sample from P, but instead, we bound the
+    // distance between the points in the base as a fraction of the diameter.
+    number_of_trials_ =
+            static_cast<int>(first_estimation * (P_diameter_ / kDiameterFraction) /
+                             max_base_diameter_);
+    if (options_.terminate_threshold < 0)
+        options_.terminate_threshold = options_.overlap_estimation;
+    if (number_of_trials_ < kMinNumberOfTrials)
+        number_of_trials_ = kMinNumberOfTrials;
+
+    printf("norm_max_dist: %f\n", options_.delta);
+    current_trial_ = 0;
+    best_LCP_ = 0.0;
+
+    Q_copy_ = Q;
+    for (int i = 0; i < 4; ++i) {
+        base_[i] = 0;
+        current_congruent_[i] = 0;
+    }
+    transform_ = Eigen::Matrix<Scalar, 4, 4>::Identity();
+
+    // call Virtual handler
+    Initialize(P, Q);
+
+    best_LCP_ = Verify(transform_);
+    printf("Initial LCP: %f\n", best_LCP_);
+}
+
+
+bool Match4PCSBase::SelectRandomTriangle(int* base1, int* base2, int* base3) {
+      if (base1 == NULL || base2 == NULL || base3 == NULL) return false;
+
+      int number_of_points = sampled_P_3D_.size();
+      *base1 = *base2 = *base3 = -1;
+
+      // Pick the first point at random.
+      int first_point = rand() % number_of_points;
+
+      // Try fixed number of times retaining the best other two.
+      Scalar best_wide = 0.0;
+      for (int i = 0; i < kNumberOfDiameterTrials; ++i) {
+        // Pick and compute
+        int second_point = rand() % number_of_points;
+        int third_point = rand() % number_of_points;
+        cv::Point3d u = sampled_P_3D_[second_point] - sampled_P_3D_[first_point];
+        cv::Point3d w = sampled_P_3D_[third_point] - sampled_P_3D_[first_point];
+        // We try to have wide triangles but still not too large.
+        Scalar how_wide = cv::norm(u.cross(w));
+        if (how_wide > best_wide && cv::norm(u) < max_base_diameter_ &&
+            cv::norm(w) < max_base_diameter_) {
+          best_wide = how_wide;
+          *base1 = first_point;
+          *base2 = second_point;
+          *base3 = third_point;
+        }
+      }
+      if (*base1 == -1 || *base2 == -1 || *base3 == -1)
+        return false;
+      else
+        return true;
+}
+
+
+
+// Try the current base in P and obtain the best pairing, i.e. the one that
+// gives the smaller distance between the two closest points. The invariants
+// corresponding the the base pairing are computed.
+bool Match4PCSBase::TryQuadrilateral(Scalar *invariant1, Scalar *invariant2,
+                                     int& id1, int& id2, int& id3, int& id4) {
+  if (invariant1 == NULL || invariant2 == NULL) return false;
+
+  float min_distance = FLT_MAX;
+  int best1, best2, best3, best4;
```

Comment:
it's better to initialize them, to -1 maybe.
and at the end of the loop put some asserts just to be sure.